src/main/scala/akka/contrib/process/NonBlockingProcess.scala

```diff
@@ -0,0 +1,270 @@
+
+    override def createLogicAndMaterializedValue(inheritedAttributes: Attributes): (GraphStageLogic, PublishIfAvailableSideChannel[T]) = {
+      val asyncCallback = new AtomicReference[AsyncCallback[PublishIfAvailableSideChannel.AsyncEvents]]
+      var downstreamReady = false
+      val downstreamReadyLock = new ReentrantLock
+      val logic = new GraphStageLogic(shape) {
+
+        override def preStart(): Unit =
+          asyncCallback.set(getAsyncCallback[PublishIfAvailableSideChannel.AsyncEvents] {
+            case PublishIfAvailableSideChannel.Publish(e: T @unchecked) =>
+              push(out, e)
+            case PublishIfAvailableSideChannel.Complete(e: Option[T] @unchecked) =>
+              e.foreach(emit(out, _))
+              completeStage()
+          })
+
+        setHandler(out, new OutHandler {
+          override def onPull(): Unit = {
+            downstreamReadyLock.lock()
+            try {
+              downstreamReady = true
+            } finally {
+              downstreamReadyLock.unlock()
+            }
+          }
+        })
+
+      }
+
+      logic -> new PublishIfAvailableSideChannel[T] {
+        override def publishIfAvailable(e: () => T): Unit =
+          Option(asyncCallback.get).foreach { ac =>
+            downstreamReadyLock.lock()
+            try {
+              if (downstreamReady) {
+                ac.invoke(PublishIfAvailableSideChannel.Publish(e()))
+                downstreamReady = false
+              }
+            } finally {
+              downstreamReadyLock.unlock()
+            }
+          }
+
+        override def complete(e: Option[T]): Unit =
+          Option(asyncCallback.get).foreach(_.invoke(PublishIfAvailableSideChannel.Complete(e)))
+      }
+    }
+  }
+
+  // For additional process detection for platforms that support "/proc"
+  private[process] val procDir = Paths.get("/proc")
+  private[process] val hasProcDir = procDir.toFile.exists()
+  // For additional checking on whether a process is alive
+  private[process] case object Inspect
+}
+
+/**
+ * NonBlockingProcess encapsulates an operating system process and its ability to be communicated with via stdio i.e.
+ * stdin, stdout and stderr. The reactive streams for stdio are communicated in a NonBlockingProcess.Started event
+ * upon the actor being established. The parent actor is then subsequently streamed
+ * stdout and stderr events. When the process exists (determined by periodically polling process.isAlive()) then
+ * the process's exit code is communicated to the receiver in a NonBlockingProcess.Exited event.
+ *
+ * All IO is performed in a non-blocking manner. Herein lies the difference between [[BlockingProcess]] and this
+ * actor. As such, many more processes can be managed via this while consuming *much* less memory on your JVM.
+ * Always favour this actor over its blocking counterpart.
+ */
+class NonBlockingProcess(
+  command: immutable.Seq[String],
+  directory: File,
```

Comment:
Should we take a `Path` instead and thus fully be using `java.nio` for everything?